### PR TITLE
#164198179 Implement user authentication with JWT

### PR DESCRIPTION
**Pull request by devdbrandy**
#### What does this PR do?

Introduce user authentication feature with JWT

#### Description of Task to be completed

- [x] Add endpoint for registration - `/api/v1/users`
- [x] Add endpoint for login - `/api/v1/users/login`
- [x] Add controller for handling authentication logic
- [x] Add middleware helper function for verifying JWT token
- [x] Write test against implementation
- [x] Update `Procfile` to temporarily run command `npm run start:dev` on heroku
- [x] Removed redundant codes and files

#### How should this be manually tested

_Prerequisite: Postgres_

1. Run the following commands on terminal:

```bash
git clone -b feature/164198179/user-authentication --single-branch \
https://github.com/andela/apollo-ah-backend.git

cd apollo-ah-backend
cp .env.example .env
npm install
```

2. Modify `.env` file with your db credentials

3. Spin up the server by running `npm run start:dev` on your terminal

4. Test endpoints `localhost:3000/api/v1/users` and `localhost:3000/api/v1/users/login` respectively with [Postman](https://www.getpostman.com/downloads/).

❗️ **Important:** Provide user payload when making request, sample below:
```json
{
  "email": "dbrandy@email.com",
  "password": "secret"
}
```

#### Any background context you want to provide

n/a

#### What are the relevant pivotal tracker stories

[#164198179](https://www.pivotaltracker.com/story/show/164198179)

#### Screenshots

n/a

### CODE REVIEW

File: test/middleware/authentication.test.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import sinon from 'sinon';
+import chai from 'chai';
+import authenticate from '../../middleware/authenticate';
+
+const should = chai.should();
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply test/middleware/authentication.test.js]: Its perfectly okay.

File: routes/users.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import express from 'express';
+import UsersController from '../controllers/users';
+
+const router = express.Router();
+
+/* user registration endpoint */
```
Comment (aanchirinah) [inline routes/users.js]: #### Critical
- Swagger Documentation required

File: routes/users.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+router.post('/', UsersController.register);
+
+/* user login endpoint */
+router.post('/login', UsersController.login);
```
Comment (aanchirinah) [inline routes/users.js]: #### Critical
- Swagger documentation required

File: .eslintrc.json [hunk 4]
```diff
@@ -28,6 +28,7 @@
             "MethodDefinition": true,
             "ClassDeclaration": true
         }
-    }]
+    }],
+    "no-unused-vars": ["error", { "args": "none" }]
```
Comment (aanchirinah) [inline .eslintrc.json]: #### Nitpick
- Why did you add this rule, it's not in the original Eslint specifications on the PT board

Comment (devdbrandy) [reply .eslintrc.json]: Thanks for pointing that. I figured that Hound repeatedly warned against unused function arguments, which in most cases is not relevant. Rule was initially added to resolve eslint `no-unused-vars` report for unused function arguments, e.g Express request object, sequelize migration etc.

File: routes/users.js (hunk 2 above)
Comment (devdbrandy) [reply routes/users.js]: Documentation provided.

File: routes/users.js (hunk 3 above)
Comment (devdbrandy) [reply routes/users.js]: Documentation provided.

File: controllers/users.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
+import bcrypt from 'bcryptjs';
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply controllers/users.js]: Removed redundant code.

File: test/middlewares/authentication.test.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import sinon from 'sinon';
+import chai from 'chai';
+import authenticate from '../../middlewares/authenticate';
+
+const should = chai.should();
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply test/middlewares/authentication.test.js]: Issue resolved

File: .eslintrc.json (hunk 4 above)
Comment (aanchirinah) [reply .eslintrc.json]: You could have just applied the rule to that file instead of the whole codebase. @devdbrandy

Comment (devdbrandy) [reply .eslintrc.json]: Yeah, that works too. I'll go ahead and make adjustments, thanks.

Comment (devdbrandy) [reply .eslintrc.json]: Resolved
